Step 5

dahun-lee-daji opened the pull request:
- 늦은 시간에 PR 드려 죄송합니다.

사고의 흐름 : 하위 모듈의 data를 상위 모듈(VC)로 가져올 수 없으니, closure를 사용하여 하위 모듈의 값을 참조한다. -> step4에서 활용한 closure를 매개로 갖는 func를 사용한다.

발생했던 문제 : 
segmentedController의 valueChanged이벤트가 발생 했을 때, playersStackView.subviews 가 제거되지 않는 문제가 발생했었습니다. -> 해결 : subviews의 subviews를 삭제하여 bottom-up으로 삭제 함.

반성 할 점 : VC 내부의 코드 가독성이 매우 안 좋게 작성되었습니다. Refactoring을 시도했으나, 손도 못대겠네요...
월요일 수업에서 배운 outputView를 분리하려 했으나, 실패하고 말았습니다..

버그 : 카드 덱의 카드가 부족할경우, gameover를 출력하고, UI에는  현재 게임 style에 맞는 수의 카드 뒷면을 나오게 하려 했으나, 커다란 하나의 뒷면만 나오는 상태입니다.  

원인 분석 -> 카드 이미지는 Player.description에 의존합니다. 덱이 부족하여 gameover 일 경우, desciption이 []로 출력되므로 한 개의 카드 뒷면이 나오는 중입니다. 

단순 해결 방안 -> pokerGame의 gameStyle을 VC로 가져와, addCard2CardView() 내부 반복문에 if cardFileNames.count == 0 조건으로 예외처리한다.

dahun-lee-daji committed:
카드 이미지 추가

dahun-lee-daji committed:
topStackView, segmentControl 추가.

테스트 미진행.
PokerGame : reset() 오버로딩 추가

dahun-lee-daji committed:
UI 구조 구현

View -> segmentStackView, playersStackView

segmentStackView -> 2 SegmentController

playersStackView -> playerInfoStackView, cardStackView

cardStackView -> Cards

dahun-lee-daji committed:
cardStackView 생성 method 구현

초기 선언에 존재하던 cardStackView를 제거
필요 시에 새롭게 생성하게 변경.

dahun-lee-daji committed:
UI 생성 method를 실행하는 method 추가

가독성 확보를 위해 변경
pokerGame의 init을 segment들의 index :0 에맞게 변경

dahun-lee-daji committed:
card이미지를 UI에 출력 기능 구현

dahun-lee-daji committed:
카드 이미지 불러오기 기능 구현

Segment의 valueChanged Handler 수정
불러온 이미지와 비교를 위해 VC.printPlayerInfo() 다시 구현
updatePlayerUIs 구현

dahun-lee-daji committed:
shake이벤트 구현, reset method 수정

기존 reset 시, Deck도 리셋했으므로, 해당 내용을 수정
Shake이벤트 구현

dahun-lee-daji committed:
Update README.md

dahun-lee-daji committed:
Update README.md

godrm commented on PokerGameApp/PokerGameApp/ViewController.swift:
```diff
@@ -9,61 +9,198 @@ import UIKit
+        case 0:
+            pokerGame.reset(with: .sevenCardStud)
+        case 1:
+            pokerGame.reset(with: .fiveCardStud)
```
여기 switch-case 구문은 GameStyle.init() 형태로 제거할 수 있습니다. 
타입을 선언했으니까 타입 자체로 검증해보세요.

godrm commented on PokerGameApp/PokerGameApp/ViewController.swift:
```diff
@@ -9,61 +9,198 @@ import UIKit
+        case 2:
+            pokerGame.reset(howMany: .three)
+        case 3:
+            pokerGame.reset(howMany: .four)
```
마찬가지로 PlayerCount.init() 형태로 switch-case 구문을 제거해보세요.

godrm commented on PokerGameApp/PokerGameApp/ViewController.swift:
```diff
@@ -9,61 +9,198 @@ import UIKit
         super.viewDidAppear(false)
-        if cardStackView.subviews.count == 0 {
+        
+        if self.view.subviews.count == 0 {
```
이 부분이 왜 필요한지 모르겠습니다. 
viewDidAppear()가 여러 번 불리는 함수라면 한 번만 불리는 함수에서 처리하면 되지 않을까요?

dahun-lee-daji replied:
초기에 한 번만 불리는 함수에서 처리하기 위해 viewDidLoad에서 구현했었습니다.
하지만, optionalBindingImage()의 Catch에 present가 들어가게 되면서, viewDidAppear()에 구현을 하였고, 이로 인하여 한 번만 불리도록 조건을 달아 두었습니다!

godrm replied:
음.. 배경 화면이 보이지 않으면 에러를 모달로 띄우는건가요? 😳
그 자체가 조금 과한 동작 같습니다. 
옵셔널 에러를 안전하게 처리하는 것은 좋은데 throw 까지 할 필요는 없구요
throw가 됐다고 present로 모달창을 띄우는 것도 과한 UI입니다. 

HIG에서도 Alert나 Modal은 사용자 경험을 방해하는 것이라 앱을 정말 사용할 수 없는 수준의 에러인 경우만 표시하도록 가이드하고 있습니다. 배경 이미지가 없는 건 개발자가 실수하는 정도니까 흐름상 아래로 내려가지 않거나 정상 흐름이 아닌 것만 인지하도록 해주면 됩니다. 그게 꼭 에러창을 띄울 필요는 없습니다. 참고하세요.

dahun-lee-daji replied:
넵! 이후 미션에서는 조언 주신 내용 바탕으로 적당히 하겠습니다! 😁

godrm commented on PokerGameApp/PokerGameApp/ViewController.swift:
```diff
@@ -9,61 +9,198 @@ import UIKit
+    
+    private func gameStart() {
+        if false == pokerGame.start() {
+            print("deck is empty, gameover. reset game")
```
앱 화면을 만들기 시작했으면 print 구문은 과감하게 빼셔도 됩니다. 
필요하면 디버거 + 브레이크포인트를 사용하세요.

godrm commented on PokerGameApp/PokerGameApp/ViewController.swift:
```diff
@@ -9,61 +9,198 @@ import UIKit
-            stackView.addArrangedSubview(generatingBacksideOfCard())
-        }
+    private func makePlayerUIs() {
+        pokerGame.showParticipatnsInfo(do: { player in
+            let info = player.description.components(separatedBy: ":")
+            let name = info[0].trimmingCharacters(in: .whitespacesAndNewlines)
+            var cards = info[1].components(separatedBy: ",").map(){
+                $0.trimmingCharacters(in: .whitespacesAndNewlines)
+            }
+            cards = cards.map(){
+                $0.trimmingCharacters(in: .punctuationCharacters)
+            }
```
player를 탐색하는 메소드를 제공하는 것은 좋지만 description으로 데이터를 전달하는 구조는 좋지 않습니다.
player에도 클로저를 넘겨주면 어떨까요?

godrm commented on PokerGameApp/PokerGameApp/ViewController.swift:
```diff
@@ -9,61 +9,198 @@ import UIKit
+            label.textAlignment = NSTextAlignment.left
+            label.text = playerName
+            return label
+        }()
```
함수 내부에서 생성할 때는 빌더패턴이 오히려 읽기 어려울 수도 있습니다. 
그냥 클로저 내부에 있는 것처럼 UILabel 생성하고 속성 지정만 해도 됩니다.

godrm commented on PokerGameApp/PokerGameApp/ViewController.swift:
```diff
@@ -77,11 +214,58 @@ class ViewController: UIViewController {
+            image = try optionalBindingImage(calledCard: cardFileName)
+        }
+        catch {
+            print(error, cardFileName)
```
try-catch 로 예외 처리를 하는 경우는 catch 부분에서는 비정상 흐름 제어를 꼭 해줘야 합니다.
에러가 나는 상황이 정상과 동일하게 아래로 내려가면 image가 없는 상태로 처리가 되겠죠? 
그런 상황인지 아닌지 판단할 수가 없습니다. throw나 return 타입을 사용하는 이유는 흐름을 제어하기 위한 것도 있습니다.

godrm commented on PokerGameApp/PokerGameApp/ViewController.swift:
```diff
@@ -77,11 +214,58 @@ class ViewController: UIViewController {
+            default:
+                fileName = "card-back"
+            }
+            cardFileNames.append(fileName)
```
여기 있는 코드도 cardName 이라는 description에 의존하는 것은 좋지 않습니다. description을 속성을 설명하는 도구일 뿐이지 그 자체가 로직과 관련이 없어야 합니다. 우리가 만든 타입이니까 괜찮겠지만, 다른 객체들은 언제든지 그 속성과 내용이 바뀔 수 있는 값입니다. 
문자열을 비교하는 것보다는 값 자체로 매핑할 수 있는 게 좋습니다.

godrm replied:
이 부분은 개선이 안된 것 같네요.

godrm commented on PokerGameApp/PokerGameApp/model/Player.swift:
```diff
@@ -32,4 +32,8 @@ class Player : CustomStringConvertible {
     func printInfo(do closure : ((Player) -> Void)) {
         closure(self)
     }
+    
+    func showInfo(do closure : ((Player) -> Void)) {
+        closure(self)
```
여기 두 가지 메소드는 시그니처가 동일해서 결국 같은 일을 하는 메소드입니다. 둘 중에 하나만 있어도 되겠네요.

godrm commented on PokerGameApp/PokerGameApp/model/PokerGame.swift:
```diff
@@ -7,20 +7,35 @@
     case one = 1, two, three, four
+    
+    static var availablePlayerCount : [String] {
+        return [self.one.description, self.two.description, self.three.description, self.four.description]
```
CaseIterable 프로토콜에 대해 학습해보세요. 직접 값을 하드코딩하는 것은 지양하세요.

godrm commented on PokerGameApp/PokerGameApp/model/PokerGame.swift:
```diff
@@ -30,25 +45,49 @@ public class PokerGame {
         return true
     }
     
     func reset(with gameStyle : GameStyle, howMany playerCount : PlayerCount) {
         self.currentGameStyle = gameStyle
         self.currentPlayerCount = playerCount
         self.dealer.resetSelf()
+        self.players = Players.init(howManyPlayer: playerCount.rawValue)
+    }
+    
+    func reset(with gameStyle : GameStyle) {
+        self.currentGameStyle = gameStyle
+        self.dealer.resetSelf()
         self.players.resetSelf()
     }
     
+    func reset(howMany playerCount : PlayerCount) {
+        self.currentPlayerCount = playerCount
+        self.dealer.resetSelf()
+        self.players = Players.init(howManyPlayer: playerCount.rawValue)
+    }
```
여기 있는 reset() 함수들은 default value를 지정해주면 같은 역할이 됩니다. 참고하세요.

dahun-lee-daji replied:
함수 값을 default로 정해준다면, 고정된 값이 들어갈텐데, 
이와 같은경우, 7stud, 4명에서, 2명으로 바꿀 때, 
default value가 5stud라면 버그가 발생 할 것 같습니다 ㅜㅜ

godrm replied:
아하.. 그렇네요. 기존 값이나 nil을 활용할 수 있는 방법이 없을까요?

godrm reviewed (changes requested):
다음 피드백 내용을 확인해보시고, 개선해서 다시 PUSH만 해보세요.

dahun-lee-daji committed:
FeedBack 반영

VC :
valueChanged 이벤트 로직 변경
카드 이미지를 불러오는 closure 내부 로직 변경

PokerGame :
availablePlayerCount return 로직 변경
reset()의 매개변수 타입을 optional로 변경하여 default로 nil을 할당.
중복된 기능의 func 제거

PlayingCard :
String을 반환하는 함수 구현
cardImage를 불러오는 closure 내 사용 됨

Deck :
[String]을 반환하는 함수 구현
cardImage를 불러오는 closure 내 사용 됨

dahun-lee-daji committed:
Merge branch 'step-5' of https://github.com/dahun-lee-daji/swift-pokergameapp into step-5

dahun-lee-daji commented:
feedBack 반영 완료했습니다~

godrm commented:
피드백 내용도 커밋을 관련 주제나 내용별로 커밋을 나누면 좋겠습니다.
다 합쳐져 있으면 특정한 것만 적용하거나 되돌리기 어렵습니다.

godrm commented on PokerGameApp/PokerGameApp/ViewController.swift:
```diff
@@ -57,32 +57,20 @@ class ViewController: UIViewController {
-            break
-        }
+        
+        let calculatedSelectedGameStyleIndex = 7 - 2 * sender.selectedSegmentIndex // 7stud : rawValue = 7, 5stud : 5 so func y = -2x + 7 , x = sender.selectedSgementIndex -> {0,1}. so y= {7,5}
```
이건 좀 오버에요. 그냥 Dictionary 같은 걸로 매칭해도 됩니다. 지금은 2개 뿐이만 여러 개면 답이 없습니다.
차라리 새로운 init()을 추가하는 게 낫겠습니다.

dahun-lee-daji replied:
아 Dictionary로 연결하면 됐겠네요! 거기까지 생각이 닿지 못했네요 ㅜㅜ

godrm commented on PokerGameApp/PokerGameApp/ViewController.swift:
```diff
@@ -131,15 +118,9 @@ class ViewController: UIViewController {
+            let cards = playerDeck.toStringArray()
             
             setPlayerInfoStackView(with: name)
             createCardStackView(with: card2FileName(with: cards))
```
cards를 넘기는 것 자체를 바꿀 줄 알았는데 playerDeck.toStringArray()만 달라졌네요. 
클로저를 넘기는 방법을 다양한 사례를 찾아보고 적용해보는 연습이 필요합니다.

godrm commented on PokerGameApp/PokerGameApp/model/Deck.swift:
```diff
@@ -62,4 +62,13 @@ struct Deck : CustomStringConvertible {
+        }
+        
+        return returnArray
+    }
```
값을 가져가지 않는 게 좋다고 피드백을 준 것 같은데,
또 다른 문자열 배열을 넘기는 함수가 추가됐네요 ㅜㅜ

godrm commented on PokerGameApp/PokerGameApp/model/PlayingCard.swift:
```diff
@@ -58,4 +58,8 @@ class PlayingCard : CustomStringConvertible{
     }
+    
+    func toString() -> String {
+        return "\(suit)\(number)"
```
스위프트에서는 toString()이란 것을 안 쓰고 대신 description을 쓰는 겁니다. 
그게 CustomStringConvertible 프로토콜이구요.

godrm commented on PokerGameApp/PokerGameApp/model/PokerGame.swift:
```diff
@@ -53,40 +55,24 @@ public class PokerGame {
+        
         self.dealer.resetSelf()
-        self.players.resetSelf()
+        self.players = Players.init(howManyPlayer: currentPlayerCount.rawValue)
```
매개변수로 currentPlayerCount.rawValue 처럼 rawValue를 넘기는 부분이 있으면 currentPlayerCount을 그대로 넘길 수 있는지 고려해보세요.

godrm commented:
피드백 이후 반영에서 정확하게 반영되지 않은 부분이 아쉽습니다.
타입 활용이나 클로저 넘기는 방식은 꼭 연습해보세요. 수고하셨습니다.

dahun-lee-daji commented:
피드백 감사드립니다! ㅜㅜ
클로저 함수의 경우에는 아무리 생각해도 값을 넘기는 것 밖에 생각이 나지 않았어요...
이에 대해서 사용 예를 많이 찾아보고 공부해야 할 것 같습니다 ㅜㅜ
